commands/command_pre_push.go

```diff
@@ -137,7 +133,19 @@ func prePushRef(left, right string) {
 			os.Exit(2)
 		}
 	}
+}
+
+func filteredPointers(pointers []*lfs.WrappedPointer, filter lfs.StringSet) []*lfs.WrappedPointer {
```

Comment:
I meant more like your callback example, channels as output and still lists as input; but I see the issue is that you need 2 iterations with shared reference data which is causing the channel way to be slow in parallel because of the mutex so it's no benefit over the callback; they'd have to be serialised to avoid the mutex, and it's more complex. 

I do prefer the callback style to returning new lists but that's probably the grumpy old man in me seeing 30% more allocations and freaking out.